### PR TITLE
refactory ShardsLimitAllocationDecider.java

### PR DESCRIPTION
**Pull request by makeyang**
re-pull request and try to refacotry code according to this: #21641 (comment)

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java [hunk 1]
```diff
@@ -185,11 +157,11 @@ public Decision canAllocate(RoutingNode node, RoutingAllocation allocation) {
...
-                "too many shards [%d] allocated to this node, [%s=%d]",
-                nodeShardCount, CLUSTER_TOTAL_SHARDS_PER_NODE_SETTING.getKey(), clusterShardLimit);
+                    "too many shards [%d] allocated to this node, [%s=%d]",
+                    nodeShardCount, CLUSTER_TOTAL_SHARDS_PER_NODE_SETTING.getKey(), clusterShardLimit);
         }
         return allocation.decision(Decision.YES, NAME,
-            "the shard count [%d] for this node is under the cluster level node limit [%d]",
-            nodeShardCount, clusterShardLimit);
+                "the shard count [%d] for this node is under the cluster level node limit [%d]",
+                nodeShardCount, clusterShardLimit);
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java]: can you undo all indentation changes, it adds noise to the diff

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java [hunk 2]
```diff
@@ -83,46 +85,17 @@ private void setClusterShardLimit(int clusterShardLimit) {
...
-        }
-        if (indexShardLimit > 0 && indexShardCount >= indexShardLimit) {
-            return allocation.decision(Decision.NO, NAME,
-                "too many shards [%d] allocated to this node for index [%s], [%s=%d]",
-                indexShardCount, shardRouting.getIndexName(), INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey(), indexShardLimit);
-        }
-        return allocation.decision(Decision.YES, NAME,
-                "the shard count [%d] for this node is under the index limit [%d] and cluster level node limit [%d]",
-                nodeShardCount, indexShardLimit, clusterShardLimit);
+        return doDecide(shardRouting, node, allocation, (count, limit) -> { return count >= limit;});
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java]: I think `(count, limit) -> count >= limit` would work too, and be slightly simpler?

**Review by jpountz — COMMENTED** [review]
I left minor comments but it looks good. @jasontedor Would you mind having a look too since you commented on the original issue?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java [hunk 3]
```diff
@@ -83,46 +85,17 @@ private void setClusterShardLimit(int clusterShardLimit) {
...
     }
 
     @Override
     public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return doDecide(shardRouting, node, allocation, (count, limit) -> {return count > limit;});
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java]: This predicate can be simplified to `(count, limit) -> count > limit`.

**Review by jasontedor — CHANGES_REQUESTED** [review]
I agree with @jpountz review, and left another comment.

**Comment by jasontedor** [comment]
Thanks @makeyang; I see that you pushed a commit incorporating our recommendations, except @jpountz was looking for you to revert all of the changes that are whitespace only changes.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java (hunk 2 above)
Comment (makeyang) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java]: already changed.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java (hunk 1 above)
Comment (makeyang) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java]: already changed.

**Comment by makeyang** [comment]
@jasontedor  @jpountz  have already changed according to u guys comments. please help to review, thanks

**Comment by jasontedor** [comment]
ok to test

**Comment by makeyang** [comment]
@jasontedor  anything I can do for testing?

**Review by jpountz — APPROVED** [review]
I'll let @jasontedor have a look too but it looks good to me now!

**Comment by jpountz** [comment]
@makeyang regarding testing, CI got all tests passing for this PR so I think we're good! https://elasticsearch-ci.elastic.co/job/elastic-elasticsearch-pull-request/1535/

**Comment by jasontedor** [comment]
@makeyang "test this please" was a command to run this PR through our CI. :smile:

The change looks good, I'll approve this PR and merge it later.

**Review by jasontedor — APPROVED** [review]
LGTM. Thanks @makeyang.

**Comment by jasontedor** [comment]
Thank you @makeyang!

**Pull request merged**